execution-engine/src/lib.rs

```diff
@@ -26,7 +26,52 @@ extern crate sgx_tstd as std;
 extern crate std;
 
 #[macro_use]
-extern crate lazy_static;
+extern crate num_derive;
 
-pub mod factory;
-pub mod hcall;
+pub mod fs;
+mod wasi;
+// Expose the error to the external.
+pub use wasi::common::FatalEngineError;
+
+#[cfg(feature = "std")]
+use crate::wasi::wasmtime::WasmtimeRuntimeState;
+use crate::{
+    fs::FileSystem,
+    wasi::{common::ExecutionEngine, wasmi::WASMIRuntimeState},
+};
+#[cfg(any(feature = "std", feature = "tz", feature = "nitro"))]
+use std::sync::Mutex;
+#[cfg(feature = "sgx")]
+use std::sync::SgxMutex as Mutex;
+use std::{boxed::Box, sync::Arc, string::ToString};
+use veracruz_utils::policy::principal::ExecutionStrategy;
+
+/// The top-level function executes program `program_name` on
+/// the `filesystem` handler, in which inputs, outputs and programs are stored.
+/// The function requires execution `strategy`.
+/// It currently supports `interp` or `JIT`, backed by `WASI` and `wasmtime`, respectively.
+/// Note that the `execute` function is essentially this library's
+/// interface to the outside world, and details exactly what external clients
+/// such as `freestanding-executuon-engine` and `runtime-manager` can rely on.
+pub fn execute(
+    strategy: &ExecutionStrategy,
+    filesystem: Arc<Mutex<FileSystem>>,
+    program_name: &str,
+) -> Result<wasi_types::ErrNo, FatalEngineError> {
+    let mut engine: Box<dyn ExecutionEngine> = match strategy {
+        ExecutionStrategy::Interpretation => {
+            Box::new(WASMIRuntimeState::new(filesystem, program_name.to_string()))
+        }
+        ExecutionStrategy::JIT => {
+            #[cfg(feature = "std")]
+            {
+                Box::new(WasmtimeRuntimeState::new(filesystem, program_name.to_string())?)
+            }
+            #[cfg(any(feature = "tz", feature = "sgx", feature = "nitro"))]
+            {
+                return Err(FatalEngineError::EngineIsNotReady);
+            }
```

Comment:
An idiomatic way to express and enforce mutual exclusivity of these blocks could be [`cfg_if`](https://docs.rs/cfg-if/1.0.0/cfg_if/).